// uu_appstatusg01/src/components/cmd-data-counter.js
"use strict";

const ResponseCounter = require("./response-counter");
const ResponseMap = require("./response-map.js");

class CmdDataCounter {
  constructor() {
    this.reset();
  }

  reset() {
    this._totalCalls = new ResponseCounter();
    this._cmdCallsMap = new ResponseMap();
    this._awidCallsMap = new ResponseMap();
    this._uuAppErrorCallsMap = {};
    this._httpStatusCodeMap = {};
  }

  record(resData) {
    this._totalCalls.record(resData.statusCode, resData.reqTime);
    if (resData.ucName) {
      this._cmdCallsMap.get(resData.ucName).record(resData.statusCode, resData.reqTime);
    }
    if (resData.awid) {
      this._awidCallsMap.get(resData.awid).record(resData.statusCode, resData.reqTime);
    }
    if (resData.errorCode) {
      this._uuAppErrorCallsMap[resData.errorCode] = (this._uuAppErrorCallsMap[resData.errorCode] || 0) + 1;
    }
    this._httpStatusCodeMap[resData.statusCode] = (this._httpStatusCodeMap[resData.statusCode] || 0) + 1;
  }

  getData() {
    return {
      totalCalls: this._totalCalls.getData(),
      uuCmdMap: this._cmdCallsMap.getData(),
      awidMap: this._awidCallsMap.getData(),
      uuAppErrorCallsMap: this._uuAppErrorCallsMap,
      httpStatusCodeMap: this._httpStatusCodeMap,
    };
  }
}

module.exports = new CmdDataCounter();
