// uu_appg01_core-validation/src/matchers/sys-matcher.js
"use strict";

const Matcher = require("../matcher.js");
const ValidationResult = require("../validation-result.js");
const Validator = require("../validator.js");

const SYS_TYPE = `const sysType = shape({
  cts: dateTime().isRequired(),
  mts: dateTime().isRequired(),
  rev: integer().isRequired(),
  lockHash: base64(),
  lockExpTs: datetime()
});`;

class _Sys extends Matcher {

  constructor(...params) {
    super("sys", ...params);
    this.validator = new Validator(SYS_TYPE);
  }

  match(value, ctx) {
    let result = new ValidationResult();
    if (value == null) {
      return result;
    }

    let childResult = this.validator.validate("sysType", value);
    if (!childResult.isValid()) {
      result.addInvalidValueError("sys.e001", `The value must be sys.`);
      Object.keys(childResult.childResults).forEach(childResultKey => {
        result.addChildResult(childResultKey, childResult.childResults[childResultKey]);
      })
    }

    return result;
  }

}

function Sys(...params) {
  let matcher = new _Sys(...params);
  if (this && this.matchers && this.matchers[0]) {
    this.matchers.push(matcher);
    return this;
  } else {
    let validator = new Validator(null);
    validator._loadMatchers();
    validator.matchers = [matcher];
    return validator;
  }
}

module.exports = Sys;
