// uu_appg01_core-validation/src/matchers/uu-app-error-map-matcher.js
"use strict";

const Matcher = require("../matcher.js");
const ValidationResult = require("../validation-result.js");
const Validator = require("../validator.js");

const UU_APP_ERROR_MAP_TYPE = `const uuAppErrorMapType = map(string(), shape({
  type: oneOf("warning", "error"),
  message: string(),
  paramMap: map(string(), any()),
  cause: shape({}, true),
}));`;

class _UuAppErrorMap extends Matcher {

  constructor(...params) {
    super("uuAppErrorMap", ...params);
    this.validator = new Validator(UU_APP_ERROR_MAP_TYPE);
  }

  match(value, ctx) {
    let result = new ValidationResult();
    if (value == null) {
      return result;
    }

    let childResult = this.validator.validate("uuAppErrorMapType", value);
    if (!childResult.isValid()) {
      result.addInvalidValueError("uuAppErrorMap.e001", `The value must be uuAppErrorMap.`);
      Object.keys(childResult.childResults).forEach(childResultKey => {
        result.addChildResult(childResultKey, childResult.childResults[childResultKey]);
      })
    }

    return result;
  }

}

function UuAppErrorMap(...params) {
  let matcher = new _UuAppErrorMap(...params);
  if (this && this.matchers && this.matchers[0]) {
    this.matchers.push(matcher);
    return this;
  } else {
    let validator = new Validator(null);
    validator._loadMatchers();
    validator.matchers = [matcher];
    return validator;
  }
}

module.exports = UuAppErrorMap;
